WeakAuras/WeakAuras.lua

```diff
@@ -2605,6 +2639,61 @@ function WeakAuras.ResolveCollisions(onFinished)
   end
 end
 
+StaticPopupDialogs["WEAKAURAS_CONFIRM_REPAIR"] = {
+  text = "",
+  button1 = L["Run the repair tool"],
+  button2 = L["Cancel"],
+  OnAccept = function(self)
+     WeakAuras.RepairDatabase()
+  end,
+  OnShow = function(self)
+    if self.data.reason == "user" then
+      self.text:SetText(L["Manual Repair Confirmation Dialog"]:format(WeakAuras.LastUpgrade()))
+    else
+      self.text:SetText(L["Automatic Repair Confirmation Dialog"]:format(WeakAuras.LastUpgrade()))
+      self.button2:SetText(L["Continue Without Repairing"])
+    end
+  end,
+  OnCancel = function(self)
+    if self.data.reason ~= "user" then
+      WeakAuras.Login()
+    end
+  end,
+  whileDead = true,
+  showAlert = true,
+  timeout = 0,
+  preferredindex = STATICPOPUP_NUMDIALOGS
+}
+
+function WeakAuras.LastUpgrade()
+  return db.lastUpgrade and date(nil, db.lastUpgrade) or "unknown"
+end
+
+function WeakAuras.NeedToRepairDatabase()
+  return db.dbVersion and db.dbVersion > WeakAuras.InternalVersion()
+end
+
+function WeakAuras.RepairDatabase(loginAfter)
+  local coro = coroutine.create(function()
+    WeakAuras.SetImporting(true)
+    -- set db version to current code version
+    db.dbVersion = WeakAuras.InternalVersion()
+    db.lastUpgrade = time()
+    -- reinstall snapshots from history
+    for id, data in pairs(db.displays) do
+      local snapshot = WeakAuras.GetMigrationSnapshot(data.uid)
+      if snapshot then
+        db.displays[id] = CopyTable(snapshot)
+        coroutine.yield()
+      end
+    end
+    WeakAuras.SetImporting(false)
+    -- finally, login
+    WeakAuras.Login()
```

Comment:
This can lead to a double call to Login(). E.g. the user log ons, gets a warning messaging but clicks on cancel. That leads to Login(). Then the user calls /wa repair, okays the repair process, which calls Login() here.

I think if the we are already logged in, we should reload the ui. (I think there's a way for addons to trigger that.)

Reply:
I tried that, it doesn't work if we use coroutines Double logins don't seem to have ill effects as best as i can tell.